enclave/core/hostcalls.c

```diff
@@ -13,6 +13,40 @@
 
 #include "core_t.h"
 
+#if !defined(OE_USE_BUILTIN_EDL)
+/**
+ * Declare the protoype of the following functions to avoid the
+ * missing-prototypes warning.
+ */
+oe_result_t _oe_log_ocall(uint32_t log_level, const char* message);
+
+oe_result_t _oe_write_ocall(int device, const char* str, size_t maxlen);
+
+/**
+ * Make the following OCALLs weak to support the system EDL opt-in.
+ * When the user does not opt in (import) the EDL, the linker will picked
+ * the following default implementations. If the user opts in the EDL,
+ * the implementions (which are strong) in the oeedger8r-generated code will be
+ * used.
+ */
+oe_result_t _oe_log_ocall(uint32_t log_level, const char* message)
+{
+    OE_UNUSED(log_level);
+    OE_UNUSED(message);
+    return OE_UNSUPPORTED;
+}
+OE_WEAK_ALIAS(_oe_log_ocall, oe_log_ocall);
+
+oe_result_t _oe_write_ocall(int device, const char* str, size_t maxlen)
```

Comment:
One interesting scenario is what would happen if one enclave supports logging and another enclave does not support logging. Since the weak-symbol will be overwritten by the strong symbol, does this mean that both the enclaves will now have logging?

Reply:
Good point. That may be the case for ecall and I need to think more about that.
For the ocall, this is fine because the enclave is compiled separately.
For example,

`enclave_foo` (does not import logging.edl) --- `_oe_write_ocall` will be picked by linker, which does nothing.
`enclave_bar` (imports the logging.edl) --- `oe_write_ocall` defined in the `_t.c` will be picked.